handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java

```diff
@@ -815,4 +816,26 @@ private static long newBIO(ByteBuf buffer) throws Exception {
             buffer.release();
         }
     }
+
+    /**
+     * Returns the {@link OpenSslKeyMaterialProvider} that should be used for OpenSSL. Depending on the given
+     * {@link KeyManagerFactory} this may cache the {@link OpenSslKeyMaterial} for better performance if it can
+     * ensure that the same material is always returned for the same alias.
+     */
+    static OpenSslKeyMaterialProvider providerFor(KeyManagerFactory factory, String password) {
+        X509KeyManager keyManager = ReferenceCountedOpenSslContext.chooseX509KeyManager(factory.getKeyManagers());
+        if (factory instanceof OpenSslCachingX509KeyManagerFactory) {
+            // The user explicit used OpenSslCachingX509KeyManagerFactory which signals us that its fine to cache.
+            return new OpenSslCachingKeyMaterialProvider(keyManager, password);
+        } else {
+            // Try to detect if its safe to cache the material.
+            String className = keyManager.getClass().getName();
+            if (className.equals("sun.security.ssl.X509KeyManagerImpl") ||
```

Comment:
I wonder if it is better just to just rely upon the user to wrap with `OpenSslCachingX509KeyManagerFactory` instead of trying to infer based upon class type. Below is the javadoc on `X509KeyManagerImpl`, which is concerning:

>   it is explicitly designed to accommodate KeyStores that change over the lifetime of the process.

If a user isn't taking advantage of this behavior then they can manually wrap in `OpenSslCachingX509KeyManagerFactory`.

Reply:
@Scottmitch yeah.... I guess better safe then sorry.